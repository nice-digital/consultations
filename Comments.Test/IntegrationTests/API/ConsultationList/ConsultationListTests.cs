using System.Collections.Generic;
using System.Linq;
using System.Net;
using Comments.Configuration;
using Comments.Test.Infrastructure;
using NICE.Feeds.Tests.Infrastructure;
using System.Threading.Tasks;
using Comments.Models;
using Comments.ViewModels;
using Newtonsoft.Json;
using Shouldly;
using Xunit;
using TestBase = Comments.Test.Infrastructure.TestBase;

namespace Comments.Test.IntegrationTests.API.ConsultationList
{
	public class ConsultationListTests : TestBase
    {
	    public ConsultationListTests() : base(TestUserType.Administrator, Feed.ConsultationCommentsListMultiple,
		    new List<SubmittedCommentsAndAnswerCount>
		    {
			    new SubmittedCommentsAndAnswerCount
			    {
				    SourceURI = "consultations://./consultation/1",
				    TotalCount = 1
			    }
			},
			submittedToLeadCommentAndAnswerCount: new List<SubmittedCommentsAndAnswerCount>
			{
				new SubmittedCommentsAndAnswerCount
				{
					SourceURI = "consultations://./consultation/1",
					OrganisationId = 1,
					TotalCount = 1
				}
			})
	    {
			AppSettings.ConsultationListConfig = TestAppSettings.GetConsultationListConfig();
		    AppSettings.Feed = TestAppSettings.GetFeedConfig();
		}

        [Fact]
        public async Task Get_Consultation_Feed_Returns_Populated_Feed_For_Administrators()
        {
			//Arrange

			// Act
			var response = await _client.GetAsync("/consultations/api/ConsultationList?Status=Open");
            response.EnsureSuccessStatusCode();
            var responseString = await response.Content.ReadAsStringAsync();

            // Assert
            responseString.ShouldMatchApproved();
        }
	}

	public class ConsultationListNotAuthTests : TestBase
	{
		public ConsultationListNotAuthTests() : base(TestUserType.NotAuthenticated, Feed.ConsultationCommentsListMultiple)
		{
			AppSettings.ConsultationListConfig = TestAppSettings.GetConsultationListConfig();
			AppSettings.Feed = TestAppSettings.GetFeedConfig();
		}

		[Fact]
		public async Task Get_Consultation_Feed_Returns_Unauthorized_for_unauthorised_users()
		{
			//Arrange in constructor

			// Act 
			var response = await _client.GetAsync("/consultations/api/ConsultationList?Status=Open");

			//Assert
			response.StatusCode.ShouldBe(HttpStatusCode.Unauthorized);
		}
	}

	public class ConsultationListAuthTests : TestBase
	{
		public ConsultationListAuthTests() : base(TestUserType.Authenticated, Feed.ConsultationCommentsListMultiple, bypassAuthentication: false)
		{
			AppSettings.ConsultationListConfig = TestAppSettings.GetConsultationListConfig();
			AppSettings.Feed = TestAppSettings.GetFeedConfig();
		}

		[Fact]
		public async Task Get_Consultation_Feed_Returns_Populated_Feed_For_Authenticated_User()
		{
			//Arrange

			// Act
			var response = await _client.GetAsync("/consultations/api/ConsultationList?Status=Open");
			response.EnsureSuccessStatusCode();
			var responseString = await response.Content.ReadAsStringAsync();

			//Assert
			responseString.ShouldMatchApproved();
		}
	}

	public class ConsultationListAuthAdminTests : TestBase
	{
		public ConsultationListAuthAdminTests() : base(TestUserType.Administrator, Feed.ConsultationCommentsListMultiple, bypassAuthentication: false, submittedCommentsAndAnswerCounts: new List<SubmittedCommentsAndAnswerCount>(){ new SubmittedCommentsAndAnswerCount(){AnswerCount = 1,CommentCount = 1,SourceURI = "source uri", TotalCount = 2}})
		{
			AppSettings.ConsultationListConfig = TestAppSettings.GetConsultationListConfig();
			AppSettings.Feed = TestAppSettings.GetFeedConfig();
		}

		[Fact]
		public async Task Get_Consultation_Feed_Returns_Populated_Feed_For_Admin_User()
		{
			//Arrange

			// Act
			var response = await _client.GetAsync("/consultations/api/ConsultationList?Status=Open");
			response.EnsureSuccessStatusCode();
			var responseString = await response.Content.ReadAsStringAsync();
			var model = JsonConvert.DeserializeObject<ConsultationListViewModel>(responseString);

			//Assert
			model.OptionFilters.Single().Options.Count(option => option.Id.Equals("Upcoming")).ShouldBe(1); //admins can see upcoming
			responseString.ShouldMatchApproved();
		}
	}

	public class ConsultationListAuthWithNoRolesTests : TestBase
	{
		public ConsultationListAuthWithNoRolesTests() : base(TestUserType.Authenticated, Feed.ConsultationCommentsListMultiple, bypassAuthentication: false, addRoleClaim: false)
		{
			AppSettings.ConsultationListConfig = TestAppSettings.GetConsultationListConfig();
			AppSettings.Feed = TestAppSettings.GetFeedConfig();
		}

		[Fact]
		public async Task Get_Consultation_Feed_Returns_Populated_Feed_For_Authenticated_User_With_No_Role()
		{
			//Arrange

			// Act
			var response = await _client.GetAsync("/consultations/api/ConsultationList?Status=Open");
			response.EnsureSuccessStatusCode();
			var responseString = await response.Content.ReadAsStringAsync();
			var model = JsonConvert.DeserializeObject<ConsultationListViewModel>(responseString);

			//Assert
			model.OptionFilters.Single().Options.Count(option => option.Id.Equals("Upcoming")).ShouldBe(0); //regular authenticated users can't see upcoming
			responseString.ShouldMatchApproved();
		}
	}
}
