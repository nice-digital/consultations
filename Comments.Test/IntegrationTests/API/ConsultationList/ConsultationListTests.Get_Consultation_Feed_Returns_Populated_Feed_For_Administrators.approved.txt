{
    "indevBasePath":"https://indevnice.org",
    "consultations":[
        {
            "title":"For consultation comments",
            "startDate":"2018-04-01T00:00:00",
            "endDate":"2028-05-31T17:00:00",
            "submissionCount":1,
            "consultationId":1,
            "documentId":1,
            "chapterSlug":"introduction",
            "gidReference":"GID-TA10232",
            "productTypeName":"Technology appraisal guidance",
            "hasCurrentUserEnteredCommentsOrAnsweredQuestions":false,
            "hasCurrentUserSubmittedCommentsOrAnswers":false,
            "allowedRole":"CHTETeam",
            "isOpen":true,
            "isClosed":false,
            "isUpcoming":false,
            "show":true,
            "organisationCodes":[
                
            ],
            "showShareWithOrganisationButton":true,
            "submissionToLeadCount":null,
            "hidden":false
        },
        {
            "title":"Transcutaneous microwave ablation for severe primary axillary hyperhidrosis ",
            "startDate":"2018-02-28T00:00:00",
            "endDate":"2028-03-23T00:00:00",
            "submissionCount":0,
            "consultationId":2,
            "documentId":2,
            "chapterSlug":"guidance",
            "gidReference":"GID-IPG10038",
            "productTypeName":"Interventional procedures guidance",
            "hasCurrentUserEnteredCommentsOrAnsweredQuestions":false,
            "hasCurrentUserSubmittedCommentsOrAnswers":false,
            "allowedRole":"CHTETeam",
            "isOpen":true,
            "isClosed":false,
            "isUpcoming":false,
            "show":true,
            "organisationCodes":[
                
            ],
            "showShareWithOrganisationButton":true,
            "submissionToLeadCount":null,
            "hidden":false
        }
    ],
    "optionFilters":[
        {
            "options":[
                {
                    "id":"Open",
                    "label":"Open",
                    "isSelected":true,
                    "filteredResultCount":2,
                    "unfilteredResultCount":2
                },
                {
                    "id":"Closed",
                    "label":"Closed",
                    "isSelected":false,
                    "filteredResultCount":0,
                    "unfilteredResultCount":2
                },
                {
                    "id":"Upcoming",
                    "label":"Upcoming",
                    "isSelected":false,
                    "filteredResultCount":0,
                    "unfilteredResultCount":2
                }
            ],
            "id":"Status",
            "title":"Status"
        }
    ],
    "textFilter":{
        "filterText":null,
        "isSelected":false,
        "filteredResultCount":2,
        "unfilteredResultCount":2,
        "id":"Keyword",
        "title":"Keyword"
    },
    "contributionFilter":[
        {
            "options":[
                {
                    "id":"HasContributed",
                    "label":"Your responses",
                    "isSelected":false,
                    "filteredResultCount":0,
                    "unfilteredResultCount":2
                }
            ],
            "id":"Contribution",
            "title":"Contribution filter group"
        }
    ],
    "teamFilter":null,
    "hiddenConsultationsFilter":[
        {
            "options":[
                {
                    "id":"ShowHiddenConsultations",
                    "label":"Show hidden consultations only",
                    "isSelected":false,
                    "filteredResultCount":0,
                    "unfilteredResultCount":2
                }
            ],
            "id":"HiddenConsultations",
            "title":"Hidden Consultations"
        }
    ],
    "user":{
        "isAdminUser":true,
        "isTeamUser":false,
        "teamNames":[
            
        ],
        "authenticatedBy":1,
        "authenticationType":"Auth0",
        "isAuthenticatedByAnyMechanism":true,
        "displayName":"Benjamin Button",
        "userId":"00000000-0000-0000-0000-000000000000",
        "organisationsAssignedAsLead":[
            
        ],
        "validatedSessions":[
            
        ],
        "validatedOrganisationUserIds":[
            
        ],
        "validatedOrganisationIds":[
            
        ],
        "isAuthenticatedByAccounts":true,
        "isAuthenticatedByOrganisationCookie":false
    },
    "initialPageView":false,
    "status":[
        "Open"
    ],
    "keyword":null,
    "contribution":null,
    "team":null,
    "hiddenConsultations":null
}